Config-level dataset info

polinaeterna opened the pull request:
Will solve https://github.com/huggingface/datasets-server/issues/864

Part of https://github.com/huggingface/datasets-server/issues/735

I preserved the old name (without leading "/") for dataset-level step - `"dataset-info"` and named config-level step `"config-info"` because  `"dataset-dataset-info"` and  `"config-dataset-info"` look a bit silly to me but I don't have a strong opinion here, let me know what you think.

`config-info` result format is dict with info, `dataset-info` result format remained the same (dict[config_name, infos]), like this:
* `config-info`
```
dataset_info: {
    "description": "_DESCRIPTION",
    "citation": "_CITATION",
    "homepage": "_HOMEPAGE",
    ...
}
```
* `dataset-info`
```
dataset_info: {
"config_1": {
    "description": "_DESCRIPTION",
    "citation": "_CITATION",
    "homepage": "_HOMEPAGE",
    ...
  },
"config_2": {
    "description": "_DESCRIPTION",
    "citation": "_CITATION",
    "homepage": "_HOMEPAGE",
    ...
  },
}
```

I'm not sure why CI is broken.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/datasets-server/pr_962). All of your documentation changes will be reflected on that endpoint.

AndreaFrancis commented on services/worker/src/worker/job_runners/config_info.py:
```diff
@@ -0,0 +1,143 @@
+        return 1
+
+    def compute(self) -> CompleteJobResult:
+        if self.config is None:
```
Maybe you could also validate dataset
        if self.dataset is None:
            raise ValueError("dataset is required")

AndreaFrancis commented on services/worker/src/worker/job_runners/config_info.py:
```diff
@@ -0,0 +1,143 @@
+
+    @staticmethod
+    def get_job_runner_version() -> int:
+        return 1
```
```suggestion
        return PROCESSING_STEP_CONFIG_INFO_VERSION
```
And import it from libcommon constants

AndreaFrancis commented:
We may need to have a migration for the change of `/dataset-info` to `dataset-info`, otherwise api will not be able to find the existing cached records for the old runner.
For both: cache and queue. You could see https://github.com/huggingface/datasets-server/blob/main/jobs/mongodb_migration/src/mongodb_migration/migrations/_20230320163700_cache_first_rows_from_streaming.py and https://github.com/huggingface/datasets-server/blob/main/jobs/mongodb_migration/src/mongodb_migration/migrations/_20230320165700_queue_first_rows_from_streaming.py

severo commented:
> I preserved the old name (without leading "/") for dataset-level step - "dataset-info" and named config-level step "config-info" because "dataset-dataset-info" and "config-dataset-info" look a bit silly to me but I don't have a strong opinion here, let me know what you think.

👍

severo commented on services/worker/src/worker/job_runners/dataset_info.py:
```diff
@@ -51,24 +61,31 @@ def __init__(self, message: str, cause: Optional[BaseException] = None):
     </Tip>
     """
-    logging.info(f"get dataset_info for dataset={dataset}")
+    logging.info(f"get dataset_info for {dataset=}")
```
```suggestion
    logging.info(f"get dataset_info for {dataset}")
```

polinaeterna replied:
don't you like this syntax? i think it's convenient to use `f"{dataset=}"` instead of `f"dataset={dataset}"`) not to duplicate words.

severo replied:
Oh, I had no idea that it was a valid syntax, I thought it was a typo. Sorry!!! And thanks for teaching me that.

severo replied:
Doc is here: https://docs.python.org/3/whatsnew/3.8.html#f-strings-support-for-self-documenting-expressions-and-debugging

severo commented on services/worker/src/worker/job_runners/dataset_info.py:
```diff
@@ -51,24 +61,31 @@ def __init__(self, message: str, cause: Optional[BaseException] = None):
+    logging.info(f"get dataset_info for {dataset=}")
 
     try:
         response = get_response(kind="/parquet-and-dataset-info", dataset=dataset)
```
I think that we should rely on `config-names` instead to get the list of configs

polinaeterna replied:
should we also do this for `dataset-size`?

severo replied:
yes! you're right, it's the same issue.

We can do it in another PR if you prefer

severo commented on services/worker/src/worker/job_runners/dataset_info.py:
```diff
@@ -78,32 +95,69 @@ def compute_dataset_info_response(dataset: str) -> DatasetInfoResponse:
+                logging.debug(f"No response found in previous step for {dataset=} {config=}: 'config-info'.")
+                pending.append(
+                    PreviousJob(
+                        kind="config-size",
```
```suggestion
                        kind="config-info",
```

severo commented on services/worker/src/worker/job_runners/dataset_info.py:
```diff
@@ -78,32 +95,69 @@ def compute_dataset_info_response(dataset: str) -> DatasetInfoResponse:
+                logging.debug(f"Previous step gave an error: {response['http_status']}")
+                failed.append(
+                    PreviousJob(
+                        kind="config-size",
```
```suggestion
                        kind="config-info",
```

polinaeterna commented:
@AndreaFrancis @severo I addressed your comments but I still have CI errors:
`ERROR tests/test_queue.py::test__add_job - RuntimeError: Mongo resource is not available` in `libcommon` and chart diff doesn't seem working. what am I missing?

severo commented:
> chart diff doesn't seem working

this is because the PR is from an external branch (in your personal fork). It works only on branches in this repo.

severo commented:
> CI errors: ERROR tests/test_queue.py::test__add_job - RuntimeError: Mongo resource is not available in libcommon

I'm not sure why. I re-run and it works now (it failed again on the first re-run)

severo commented on services/worker/src/worker/job_runners/config_info.py:
I think the file should now live under job_runners/config/info.py, right @AndreaFrancis ?

severo commented on services/worker/src/worker/job_runners/dataset_info.py:
and this file should live in job_runners/dataset/info.py

severo reviewed (approved):
Good work, thanks!

Just a detail with the name of two files

AndreaFrancis commented on jobs/mongodb_migration/src/mongodb_migration/migrations/_20230323160000_queue_dataset_info.py:
```diff
@@ -0,0 +1,71 @@
+
+dataset_info = "/dataset-info"
+dataset_info_updated = "dataset-info"
+db_name = "cache"
```
```suggestion
```
It looks like this is not used

AndreaFrancis reviewed (approved):
Nice!
